experimental/inductor.py

```diff
@@ -0,0 +1,411 @@
+
+
+def train_step(m, o, x, y):
+    # if we dont set to none, the gradients will be copied into the buffers at the end
+    # that's alright but probably a perf regression
+    o.zero_grad(set_to_none=True)  
+
+    y_hat = m.forward(x)
+    loss = torch.sum((y_hat - y) ** 2)
+    loss.backward()
+    o.step()
+    return (loss, )
+
+
+def main(optim_cls):
+    x = torch.ones(10, 4, device='cuda')
+    y = torch.ones(10, 2, device='cuda')
+    m = MyModule().to('cuda')
+    sgd = optim_cls(m.parameters(), lr=0.01)
+
+    from copy import deepcopy
+    mc = deepcopy(m)
+
+    print ('==== Eager ====')
+    # eager
+    for i in range(10): 
+        loss = train_step(m, sgd, x, y)
+        print(loss)
+
+    print('==== Compiled ====')
+
+    sgd = optim_cls(mc.parameters(), lr=0.01)
+
+    if DIST:
+        mesh = DeviceMesh('cuda', range(torch.distributed.get_world_size()))
+        default_shard_schema = Schema(mesh, [Replicate()])
+        mark_placements(x, [Shard(0)])
+        mark_placements(y, [Shard(0)])
+
+        x_shard = get_example_shard(x, mesh)
+        y_shard = get_example_shard(y, mesh)
+        dprint(f'Input shape: {x.shape}; shard shape: {x_shard.shape}')
+    else:
+        x_shard = x
+        y_shard = y
+        default_shard_schema = None
+
+
+    # warning: even though we'ree passing mc and sgd as "example" inputs,
+    #    the tracer actually takes its tensor states and saves them as part of
+    #    the compiled_f. So mc and sgd cannot be mutated directly afterwards.
+    compiled_f = make_inductor_fn(train_step, [mc, sgd, x, y], default_shard_schema=default_shard_schema)
+
+    if DIST:
+        # make sure our parameter is the same across devices initially
+        # doing this after compile because during compilation we're running the training step a few times
+        # TODO: fix this. we should run compilation with fake tensors.
+        for p in mc.parameters(): 
+            torch.distributed.all_reduce(p)
+
+    with torch.no_grad():
+        for i in range(10): 
+            # FIXME attention -- mc and sgd are actualy _not_ being used!!
+            #    this is because we collected parameter/tensors from mc and sgd
+            #    during the first tracing, and we're reusing those here instead of
+            #    actually taking them from mc and sgd.
+            loss_shard,  = compiled_f(mc, sgd, x_shard, y_shard)
+            dprint(f'iteration: local_loss={loss_shard}')
+
+
+if __name__ == '__main__':
+    if DIST:
+        torch.distributed.init_process_group(backend='nccl')
+        torch.cuda.set_device(torch.distributed.get_rank() % 8)
+
+    print('================ main_simple =================')
+    main_simple()
+
+    print('================ main with SGD =================')
+    main(torch.optim.SGD)
+
+    print('================ main with Adam =================')
+    # sharding seems to be wrong -- seems to be doing a bunch of allreduces that are not needed
```

Comment:
See outputs here: 
https://gist.github.com/aazzolini/4e7b8580fe0b5cebdefe0a7307a9a00f

Notice multiple all_reduce being inserted after the forward pass